packages/account/src/providers/provider.ts

```diff
@@ -680,45 +692,42 @@ export default class Provider {
    * @param transactionRequest - The transaction request object.
    * @returns A promise.
    */
-  async estimateTxDependencies(transactionRequest: TransactionRequest): Promise<void> {
-    let missingOutputVariableCount = 0;
-    let missingOutputContractIdsCount = 0;
-    let tries = 0;
-
+  async estimateTxDependencies(transactionRequest: TransactionRequest): Promise<CallResult> {
     if (transactionRequest.type === TransactionType.Create) {
-      return;
+      return {
+        receipts: [],
+      };
     }
 
-    let txRequest = transactionRequest;
+    await this.estimatePredicates(transactionRequest);
 
-    if (txRequest.hasPredicateInput()) {
-      txRequest = (await this.estimatePredicates(txRequest)) as ScriptTransactionRequest;
-    }
+    let receipts: TransactionResultReceipt[] = [];
 
-    while (tries < MAX_RETRIES) {
+    for (let attempt = 0; attempt < MAX_RETRIES; attempt++) {
       const { dryRun: gqlReceipts } = await this.operations.dryRun({
-        encodedTransaction: hexlify(txRequest.toTransactionBytes()),
+        encodedTransaction: hexlify(transactionRequest.toTransactionBytes()),
         utxoValidation: false,
       });
-      const receipts = gqlReceipts.map(processGqlReceipt);
+      receipts = gqlReceipts.map(processGqlReceipt);
       const { missingOutputVariables, missingOutputContractIds } =
         getReceiptsWithMissingData(receipts);
 
-      missingOutputVariableCount = missingOutputVariables.length;
-      missingOutputContractIdsCount = missingOutputContractIds.length;
+      const hasMissingOutputs =
+        missingOutputVariables.length !== 0 || missingOutputContractIds.length !== 0;
 
-      if (missingOutputVariableCount === 0 && missingOutputContractIdsCount === 0) {
-        return;
+      if (!hasMissingOutputs) {
+        break;
       }
 
-      txRequest.addVariableOutputs(missingOutputVariableCount);
-
-      missingOutputContractIds.forEach(({ contractId }) =>
-        txRequest.addContractInputAndOutput(Address.fromString(contractId))
-      );
-
-      tries += 1;
+      transactionRequest.addVariableOutputs(missingOutputVariables.length);
+      missingOutputContractIds.forEach(({ contractId }) => {
+        transactionRequest.addContractInputAndOutput(Address.fromString(contractId));
+      });
     }
+
+    return {
+      receipts,
+    };
```

Comment:
IMO it would be a bit cleaner / clearer (less lines and avoid double negative) to do the following:

```ts
if (missingOutputVariables.length || missingOutputContractIds.length) {
  transactionRequest.addVariableOutputs(missingOutputVariables.length);
  missingOutputContractIds.forEach(({ contractId }) => {
    transactionRequest.addContractInputAndOutput(Address.fromString(contractId));
   });
}

break;
```